contracts/keep3r-liquidity-manager/Keep3rLiquidityManagerWork.sol

```diff
@@ -20,64 +20,118 @@ interface IKeep3rLiquidityManagerWork {
 
   function forceWork(address _job) external;
 }
 
-abstract contract Keep3rLiquidityManagerWork is
-  Keep3rLiquidityManagerEscrowsHandler,
-  Keep3rLiquidityManagerUserJobsLiquidityHandler,
-  IKeep3rLiquidityManagerWork
-{
+abstract contract Keep3rLiquidityManagerWork is Keep3rLiquidityManagerUserJobsLiquidityHandler, IKeep3rLiquidityManagerWork {
+  // job => escrow => Steps
+  mapping(address => mapping(address => Steps)) public override jobEscrowStep;
+  // job => escrow => timestamp
+  mapping(address => mapping(address => uint256)) public override jobEscrowTimestamp;
+
+  constructor(
+    address _keep3rV1,
+    address _escrow1,
+    address _escrow2
+  ) public Keep3rLiquidityManagerUserJobsLiquidityHandler(_keep3rV1, _escrow1, _escrow2) {}
+
   // Since all liquidity behaves the same, we just need to check one of them
   function getNextAction(address _job) public view override returns (address _escrow, Actions _action) {
     require(jobLiquidities[_job].length > 0, 'Keep3rLiquidityManager::getNextAction:job-has-no-liquidity');
-    address _liquidity = jobLiquidities[_job][0];
 
-    uint256 liquidityProvided1 = IKeep3rV1(keep3rV1).liquidityProvided(address(escrow1), _liquidity, _job);
-    uint256 liquidityProvided2 = IKeep3rV1(keep3rV1).liquidityProvided(address(escrow2), _liquidity, _job);
-    if (liquidityProvided1 == 0 && liquidityProvided2 == 0) {
-      // Only start if both escrow have liquidity
-      require(IERC20(_liquidity).balanceOf(address(escrow1)) > 0, 'Keep3rLiquidityManager::getNextAction:escrow1-liquidity-is-0');
-      require(IERC20(_liquidity).balanceOf(address(escrow2)) > 0, 'Keep3rLiquidityManager::getNextAction:escrow2-liquidity-is-0');
+    Steps _escrow1Step = jobEscrowStep[_job][escrow1];
+    Steps _escrow2Step = jobEscrowStep[_job][escrow2];
 
-      // Start by _addLiquidityToJob liquidity with escrow1 as default
+    // Init (add liquidity to escrow1)
+    if (_escrow1Step == Steps.NotStarted && _escrow2Step == Steps.NotStarted) {
       return (escrow1, Actions.AddLiquidityToJob);
     }
 
-    // The escrow with liquidityAmount is the one to call applyCreditToJob, the other should call unbondLiquidityFromJob
-    if (
-      IKeep3rV1(keep3rV1).liquidityAmount(address(escrow1), _liquidity, _job) > 0 &&
-      IKeep3rV1(keep3rV1).liquidityApplied(address(escrow1), _liquidity, _job) < block.timestamp
-    ) {
-      return (escrow1, Actions.ApplyCreditToJob);
+    // Init (add liquidity to NotStarted escrow)
+    if ((_escrow1Step == Steps.NotStarted || _escrow2Step == Steps.NotStarted) && _jobHasDesiredLiquidity(_job)) {
+      _escrow = _escrow1Step == Steps.NotStarted ? escrow1 : escrow2;
+      address _otherEscrow = _escrow == escrow1 ? escrow2 : escrow1;
+
+      // on _otherEscrow step CreditApplied
+      if (jobEscrowStep[_job][_otherEscrow] == Steps.CreditApplied) {
+        // make sure to wait 14 days
+        if (block.timestamp > jobEscrowTimestamp[_job][_otherEscrow].add(14 days)) {
```

Comment:
although it would make it less clear, i think combining this too can be more gas efficient (i like code being clear tho lol)

Reply:
Let's make this improvements on the last review. tight now I chose to prioritize readability.